fixed the wechat platform can't idenitfy the ttf problem

ColinCollins opened the pull request:
fixed WeChat playform can't identify the ttf font format problem 


fixed WeChat playform can't identify the ttf font format problem

fix #2456 cocos-creator/fireball#7066

Changelog:
 * 支持在微信小游戏中使用 TTF 字体

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!

- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.

- To official teams:
  - [ ] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
  - [ ] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
  - [ ] Make sure any runtime log information in `cc.log` , `cc.error` or `new Error('')` has been moved into `DebugInfos.js` with an ID, and use `cc.logID(id)` or `new Error(cc._getErrorID(id))` instead.

-->

jareguo commented on cocos2d/core/label/CCSGLabel.js:
```diff
@@ -490,8 +490,18 @@ _ccsg.Label = _ccsg.Node.extend({
+        var fontFamilyName = "";
+        if (CC_WECHATGAME) {
+            fontFamilyName = wx.loadFont(fontHandle);
+            if (fontFamilyName == null) {
```
你确定你这边要用的是两个等号吗？两个等号的意思是 fontFamilyName === null || fontFamilyName === undefined。而这种情况下，一般用 if (!fontFameilyName) 会更没歧义一点。

jareguo commented on cocos2d/core/label/CCSGLabel.js:
```diff
@@ -490,8 +490,18 @@ _ccsg.Label = _ccsg.Node.extend({
+            if (fontFamilyName == null) {
+                return cc.CustomFontLoader._getFontFamily(fontHandle);
+            }
+            if (!fontHandle in cc.CustomFontLoader._fontFamilyCache) {
```
in 运算符的优先级是小于取反运算的，所以这里肯定不能这么写

jareguo commented on cocos2d/core/label/CCTextUtils.js:
```diff
@@ -50,12 +50,20 @@ CustomFontDescriptor.prototype.addHandler = function (callback) {
 var CustomFontLoader = {
     _fontCache: {},
     _fontWidthCache: {},
+    _fontFamilyCache:{},
```
为什么要 cache 呢？冒号后加个空格哈

jareguo commented on cocos2d/core/label/CCTextUtils.js:
```diff
@@ -50,12 +50,20 @@ CustomFontDescriptor.prototype.addHandler = function (callback) {
+        if (CC_WECHATGAME) {
+            if (url in this._fontFamilyCache) {
+                fontFamilyName = this._fontFamilyCache[url];
+            }
```
else 的话会怎样？

jareguo commented on cocos2d/core/label/CCSGLabel.js:
```diff
@@ -490,8 +490,18 @@ _ccsg.Label = _ccsg.Node.extend({
+        if (CC_WECHATGAME) {
+            fontFamilyName = wx.loadFont(fontHandle);
+            if (!fontFameilyName) {
+                return cc.CustomFontLoader._getFontFamily(fontHandle);
```
为什么这里 return 了，其它分支没 return？

pandamicro replied:
下面的 callback 还是需要执行的，需要做 font 加载完成后的通知和处理

jareguo commented on cocos2d/core/label/CCSGLabel.js:
```diff
@@ -490,8 +490,18 @@ _ccsg.Label = _ccsg.Node.extend({
+                return cc.CustomFontLoader._getFontFamily(fontHandle);
+            }
+            if (!(fontHandle in cc.CustomFontLoader._fontFamilyCache)) {
+                cc.CustomFontLoader._fontFamilyCache[fontHandle] = fontFamilyName;
```
为什么只有 微信需要用 cache？

jareguo commented on cocos2d/core/label/CCSGLabel.js:
```diff
@@ -490,8 +490,18 @@ _ccsg.Label = _ccsg.Node.extend({
+            if (!fontFameilyName) {
+                return cc.CustomFontLoader._getFontFamily(fontHandle);
+            }
+            if (!(fontHandle in cc.CustomFontLoader._fontFamilyCache)) {
```
走到这里后，下面还是会执行到 CustomFontLoader.loadTTF，从代码上看，load 了两次字体啊。太奇怪了，加点注释？

pandamicro commented on cocos2d/core/label/CCTextUtils.js:
```diff
@@ -50,12 +50,22 @@ CustomFontDescriptor.prototype.addHandler = function (callback) {
+            if (url in this._fontFamilyCache) {
+                fontFamilyName = this._fontFamilyCache[url];
+            } else {
+                fontFamilyName = this._getFontFamily(url);
```
这个分支有意义吗？肯定是无法加载成功的把？

pandamicro commented on cocos2d/core/label/CCTextUtils.js:
```diff
@@ -50,12 +50,22 @@ CustomFontDescriptor.prototype.addHandler = function (callback) {
 var CustomFontLoader = {
     _fontCache: {},
     _fontWidthCache: {},
+    _fontFamilyCache: {},
```
这是只有微信小游戏平台才需要的，在命名上表现出来

jareguo reviewed (changes requested):
还是有点问题

jareguo commented on cocos2d/core/label/CCSGLabel.js:
```diff
@@ -490,13 +490,24 @@ _ccsg.Label = _ccsg.Node.extend({
+        var fontFamilyName = "";
+        if (CC_WECHATGAME) {
+            fontFamilyName = wx.loadFont(fontHandle);        
+            //avoid the error in wechat editor platform
```
wechat devtools

ColinCollins committed:
fixed WeChat playform can't identify the ttf font format problem

fixed ParticleSystem part memory  leak problem


fixed the wechat can't identify ttf format problem


s


fixed WeChat playform can't identify the ttf font format problem


fixed WeChat playform can't identify the ttf font format problem


s
